orchestrator/careplanservice/handle_createtask.go

```diff
@@ -33,47 +34,54 @@ func (s *Service) handleCreateTask(ctx context.Context, request FHIRHandlerReque
 		return nil, errors.New(fmt.Sprintf("cannot create Task with status %s, must be %s or %s", task.Status, fhir.TaskStatusRequested.String(), fhir.TaskStatusReady.String()))
 	}
 
-	var carePlan fhir.CarePlan
+	principal, err := auth.PrincipalFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	carePlan := fhir.CarePlan{}
 	var carePlanRef *string
-	err := coolfhir.ValidateTaskRequiredFields(task)
+	err = coolfhir.ValidateTaskRequiredFields(task)
 	if err != nil {
 		return nil, err
 	}
 	// Resolve the CarePlan
-	if len(task.BasedOn) == 0 {
+	if task.BasedOn == nil || len(task.BasedOn) == 0 {
 		// The CarePlan does not exist, a CarePlan and CareTeam will be created and the requester will be added as a member
-		cp := fhir.CarePlan{}
-
-		carePlanBytes, _ := json.Marshal(cp)
-		newHttpRequest := httptest.NewRequest("POST", "/CarePlan", bytes.NewReader(carePlanBytes))
-
-		// Create a new transaction, we need these resources to be created before the flow can continue
-		newTx := coolfhir.Transaction()
-
-		res, err := s.handleCreateCarePlan(newHttpRequest, newTx)
-		if err != nil {
-			return nil, err
-		}
-		if res == nil {
-			return nil, errors.New("invalid response")
-		}
-
-		bundle, err := s.commitTransaction(newHttpRequest, newTx, []FHIRHandlerResult{res})
-		if err != nil {
-			return nil, err
-		}
-
-		err = coolfhir.ResourceInBundle(bundle, coolfhir.EntryIsOfType("CarePlan"), &carePlan)
-		if err != nil {
-			return nil, err
-		}
-
+		// Create a new CarePlan (which will also create a new CareTeam) based on the Task reference
```

Comment:
this looks much better and convoluted

Reply:
Hopefully you mean "much better and *less* convoluted" :P